### PR TITLE
Emit notice logs in :console backend

### PR DESCRIPTION
**Pull request by riccardomanfrin**
yes.. this is not very "backward compatible"..
I guess adding some config flag to enable it, and
default it to false.. would be proper (?)

### CODE REVIEW

**Comment by riccardomanfrin** [comment]
I've added support for `:notice` in the `formatter.ex` and changed the `console.ex` backend.. I hope this is limiting the propagation of the `:notice` to only the console backend, leaving the rest of backends happy (the key was there already as `erl_level`)

Is this what you meant?

It's in any case breaking the format of preexisting log collections in case someone was logging to notice.

File: lib/logger/lib/logger/backends/console.ex [hunk 1]
```diff
@@ -232,15 +232,21 @@ defmodule Logger.Backends.Console do
...
       warn: Keyword.get(colors, :warn, :yellow),
       error: Keyword.get(colors, :error, :red),
+      notice: Keyword.get(colors, :error, :blue),
       enabled: Keyword.get(colors, :enabled, IO.ANSI.enabled?())
     }
   end
 
-  defp log_event(level, msg, ts, md, %{device: device} = state) do
+  defp log_event(_level, msg, ts, md, %{device: device} = state) do
+    {:erl_level, level} = List.keyfind(md, :erl_level, 0, :not_found)
```
Comment (josevalim) [inline lib/logger/lib/logger/backends/console.ex:248]:
```suggestion
    {:erl_level, level} = List.keyfind(md, :erl_level, 0, {:erl_level, :not_found})
```

File: lib/logger/lib/logger/backends/console.ex [hunk 2]
```diff
@@ -232,15 +232,21 @@ defmodule Logger.Backends.Console do
...
       warn: Keyword.get(colors, :warn, :yellow),
       error: Keyword.get(colors, :error, :red),
+      notice: Keyword.get(colors, :error, :blue),
       enabled: Keyword.get(colors, :enabled, IO.ANSI.enabled?())
     }
   end
 
-  defp log_event(level, msg, ts, md, %{device: device} = state) do
+  defp log_event(_level, msg, ts, md, %{device: device} = state) do
+    {:erl_level, level} = List.keyfind(md, :erl_level, 0, {:erl_level, :not_found})
```
[the remark this answers is not in the document]
Comment (josevalim) [reply lib/logger/lib/logger/backends/console.ex:248]: Good find!

File: lib/logger/lib/logger/backends/console.ex [hunk 3]
```diff
@@ -138,6 +138,12 @@ defmodule Logger.Backends.Console do
   def handle_event({level, _gl, {Logger, msg, ts, md}}, state) do
     %{level: log_level, ref: ref, buffer_size: buffer_size, max_buffer: max_buffer} = state
 
+    {:erl_level, level} = List.keyfind(md, :erl_level, 0, {:erl_level, level})
+    level = case level do
+      :warning -> :warn
+      _ -> level
+    end
```
Comment (josevalim) [inline lib/logger/lib/logger/backends/console.ex:145]: No, let's not convert back. `Logger.warn` is already deprecated, so it is a good opportunity to move to `warning`. :)

File: lib/logger/lib/logger/backends/console.ex [hunk 4]
```diff
@@ -243,7 +239,7 @@ defmodule Logger.Backends.Console do
       emergency: Keyword.get(colors, :debug, :red),
       debug: Keyword.get(colors, :debug, :cyan),
       info: Keyword.get(colors, :info, :normal),
-      warn: Keyword.get(colors, :warn, :yellow),
+      warning: Keyword.get(colors, :warning, :yellow),
```
Comment (josevalim) [inline lib/logger/lib/logger/backends/console.ex:242]:
```suggestion
      warning: Keyword.get(colors, :warn, :yellow),
```

File: lib/logger/test/logger/backends/console_test.exs [hunk 5]
```diff
@@ -154,7 +166,7 @@ defmodule Logger.Backends.ConsoleTest do
     Logger.configure_backend(:console, colors: [warn: :cyan])
 
     assert capture_log(fn -> Logger.warn("hello") end) ==
-             IO.ANSI.cyan() <> "hello" <> IO.ANSI.reset()
+             IO.ANSI.yellow() <> "hello" <> IO.ANSI.reset()
```
Comment (josevalim) [inline lib/logger/test/logger/backends/console_test.exs:169]:
```suggestion
             IO.ANSI.cyan() <> "hello" <> IO.ANSI.reset()
```

**Review by hauleth — CHANGES_REQUESTED** [review]
Ignoring all the comments, I have different concept how we can add support for more features from Erlang logger instead.

- add new optional callback for the backends, for example `type/0` that will return `:legacy | :rich` and will default to `:legacy` if not defined
- for `:legacy` backends receives message in the current, tuple based, format
- for `:rich` (can use different name there) backend receives message in "raw" format from the Erlang (possibly with some alterations to wrap `:report_cb` for example)

This approach should give us a way to make full transition in future (possibly still in multiple steps).

What do you think @josevalim about such approach? It would give a way to have smother transition period, and possibly migrate to "new" format even without bumping Logger's major version (which would imply bumping Elixir's major version).

File: lib/logger/lib/logger/formatter.ex [hunk 6]
```diff
@@ -163,10 +163,10 @@ defmodule Logger.Formatter do
   defp output(:levelpad, level, _, _, _), do: levelpad(level)
   defp output(other, _, _, _, _), do: other
 
-  defp levelpad(:debug), do: ""
+  # TODO: Deprecate me on Elixir v1.13+ or later
```
[the remark this answers is not in the document]
Comment (josevalim) [reply lib/logger/lib/logger/formatter.ex:166]: levelpad is not very useful now when one of the levels is "emergency". There will be just too much padding on something like "info" (5 chars to be more precise).

**Comment by hauleth** [comment]
@josevalim you probably meant #7928. I can see if I can craft something like that as a part of Hacktoberfest.

I wouldn't say that it is strictly needed to get us there, but for sure can ease transition period (unless I really will be able to craft my PR before 1.13).

**Pull request merged**